### PR TITLE
[Storage] Bucket Lock

### PR DESCRIPTION
**Pull request by frankyn**
### Samples
- [x] Add check when cleaning up bucket in case temporary or event-based holds are enabled for an object.
- [x]  Tested samples
- [x]  Define a retention policy
- [x]  Remove a retention policy if not locked
- [x]  Lock a retention policy
- [x]  Set temporary hold
- [x]  Release temporary hold
- [x]  Set event-based hold
- [x]  Release event-based hold
- [x]  Set default event-based hold
- [x]  Release default event-based hold
- [x]  View default event-based hold
- [x]  View bucket retention policy metadata
- [x]  View metadata for object

### CODE REVIEW

**Comment by frankyn** [comment]
@tbpg could you review this PR when you have a moment?

File: storage/buckets/main.go [hunk 1]
```diff
@@ -197,6 +198,133 @@ func removeUser(c *storage.Client, bucketName string) error {
...
+
+func getDefaultEventBasedHold(c *storage.Client, bucketName string) (*storage.BucketAttrs, error) {
+	ctx := context.Background()
+
+	// [START storage_get_default_event_based_hold]
+	attrs, err := c.Bucket(bucketName).Attrs(ctx)
+	if err != nil {
+		return nil, err
+	}
+	fmt.Printf("Is default event-based hold enabled? %t\n",
```
[the remark this answers is not in the document]
Comment (frankyn) [reply storage/buckets/main.go]: I'll follow-up on this when samples are staged to make sure, but I didn't want accidentally have code hidden due to length.

Thanks for calling it out.

File: storage/buckets/main.go [hunk 2]
```diff
@@ -197,6 +198,133 @@ func removeUser(c *storage.Client, bucketName string) error {
...
+	}
+	if err := bucket.If(conditions).LockRetentionPolicy(ctx); err != nil {
+		return err
+	}
+
+	lockedAttrs, err := c.Bucket(bucketName).Attrs(ctx)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Retention policy for %v is now locked\n", bucketName)
```
[the remark this answers is not in the document]
Comment (frankyn) [reply storage/buckets/main.go]: Based on how the DLP sample uses `io.Writer`, looks like I would do,
`fmt.Fprintf(w, "Retention policy for %v is now locked\n", bucketName)`
after `[END storage_lock_retention_policy]` as to not confuse sample readers. 

An alternative might be to review `attrs` returned by `getRetentionPolicy`. Is that reasonable?

**Comment by frankyn** [comment]
Agreed, I'll add in exponential backoff.

File: storage/buckets/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbpg) [reply storage/buckets/main.go]: Reviewing `attrs` sounds good to me. If we print to stdout, we should capture the output in tests like this: https://github.com/GoogleCloudPlatform/golang-samples/blob/master/monitoring/custommetric/custommetric_test.go#L18

[the remark this answers is not in the document]
Comment (frankyn) [reply storage/buckets/main.go]: Thanks! will do.

**Comment by frankyn** [comment]
scratch exp backoff, keeping it simple for now.

**Comment by frankyn** [comment]
Rebased and addressed feedback, @tbpg PTAL when you have a moment. Thank you!

File: storage/objects/main.go [hunk 3]
```diff
@@ -177,10 +177,87 @@ func attrs(client *storage.Client, bucket, object string) (*storage.ObjectAttrs,
...
+
 	return attrs, nil
 	// [END get_metadata]
 }
 
+func setEventBasedHold(client *storage.Client, bucket, object string) error {
+	ctx := context.Background()
+	// [START storage_set_event_based_hold]
+	o := client.Bucket(bucket).Object(object)
+	if _, err := o.Update(ctx, storage.ObjectAttrsToUpdate{
```
[the remark this answers is not in the document]
Comment (frankyn) [reply storage/objects/main.go]: whoops! Done.

**Comment by frankyn** [comment]
TY!

**Pull request merged**